src/auth.zig

```diff
@@ -107,6 +97,296 @@ test MTLS {
+                    extra_data = decoded[8..][i + 1 ..][0..ed];
+                    given_hash = decoded[8..][i + 1 ..][ed + 1 .. HMAC.mac_length];
+                }
+
+                @memcpy(user_buffer[0..username.len], username);
+
+                hm.update(time);
+                hm.update(username);
+                if (extra_data) |ed| hm.update(ed);
+                var our_hash: [HMAC.mac_length]u8 = undefined;
+                hm.final(our_hash[0..]);
+                if (std.crypto.utils.timingSafeEql([HMAC.mac_length]u8, given_hash[0..HMAC.mac_length].*, our_hash)) {
+                    return username;
+                }
+
+                return error.InvalidAuth;
+            } else return error.InvalidAuth;
+            return error.InvalidAuth;
+        }
+
+        pub fn authenticate(ptr: *anyopaque, headers: *const Headers) Error!User {
+            const ca: *Self = @ptrCast(@alignCast(ptr));
+            if (ca.base) |base| {
+                // If base provider offers authenticate, we should defer to it
+                if (base.vtable.authenticate) |_| {
+                    return base.authenticate(headers);
+                }
+
+                if (headers.get("Cookie")) |cookies| {
+                    // This actually isn't technically invalid, it's only
+                    // currently not implemented.
+                    if (cookies.value_list.next != null) return error.InvalidAuth;
+                    const cookie = cookies.value_list.value;
+                    std.debug.print("cookie: {s} \n", .{cookie});
+                    var itr = std.mem.tokenizeSequence(u8, cookie, "; ");
+                    while (itr.next()) |tkn| {
+                        if (startsWith(u8, tkn, ca.cookie_name)) {
+                            var un_buf: [64]u8 = undefined;
+                            var hmac = HMAC.init(ca.server_secret_key);
+                            const username = try validateToken(
+                                &hmac,
+                                tkn[ca.cookie_name.len + 1 ..],
+                                un_buf[0..],
+                            );
+
+                            return base.lookupUser(username);
+                        }
+                    }
+                } else {
+                    std.debug.print("no cookie\n{any}", .{headers});
+                }
+            }
+
+            return error.UnknownUser;
+        }
+
+        pub fn valid(ptr: *anyopaque, user: *const User) bool {
+            const ca: *Self = @ptrCast(@alignCast(ptr));
+            if (ca.base) |base| return base.valid(user);
+            return false;
+        }
+
+        pub fn lookupUser(ptr: *anyopaque, user_id: []const u8) Error!User {
+            const ca: *Self = @ptrCast(@alignCast(ptr));
+            if (ca.base) |base| return base.lookupUser(user_id);
+            return error.UnknownUser;
+        }
+
+        pub fn mkToken(hm: *HMAC, token: []u8, user: *const User) Error!usize {
```

Comment:
Does this function need to be public?

Reply:
same answer, with the bonus detail, is if/when I add an abstraction over the token generation into it's own type, it *will* need to be, and then it'll probably have a better place to live